### ISSUE TITLE
Changing the value of epsilon

### ISSUE THREAD

**Issue by mnaseri2770** [ChrisWaites/pyvacy#2]
Hi, 

Is it possible to apply DP with different values of epsilon?

Thanks,

**Comment by ChrisWaites** [comment]
Yes! The epsilon you achieve is just a function of the training parameters, i.e. what you set for clipping, number of epochs, etc. Although, if the question is whether you can set epsilon explicitly and infer the value of some other free variable, the current implementation does not easily allow for this.

**Comment by mnaseri2770** [comment]
Well, then is there anyway to modify the privacy budget? I assumed we do that by changing the epsilon value.

**Comment by ChrisWaites** [comment]
Good q! So to be clear you _can_ modify epsilon, just implicitly - if you want a lower epsilon, just change the value for one of your training parameters e.g. lower the number of epochs, increase the noise multiplier, etc. I'm just saying that the implementation doesn't allow for you explicitly set epsilon to, say, precisely 3.0.

With respect to the privacy budget, that's just a value you commit to outside the scope of this framework. I.e. this library simply allows you to calculate the epsilon associated with a given training procedure. Then, you could theoretically add up several epsilons associated with multiple training procedures, make sure they don't exceed your budget, whatever you want.

**Comment by mnaseri2770** [comment]
Alright, then maybe I misunderstood epsilon. Because still I am not sure why epsilon needs to be a function of the training parameters. I thought the defines the level of privacy that we would like to apply, and that is why it should be possible to modify it.

**Comment by ChrisWaites** [comment]
So "the level of privacy that you would like to apply" comes from the training parameters. For example, there's a parameter called the noise multiplier - that's how large the noise is that you apply to the gradients. So naturally if you set this very high, your training will be more private because you're adding more noise. Then, given all the training parameters, epsilon is a _measure_ of the privacy you've achieved, rather than something you use during training to make things more or less private. Hopefully this makes sense.

**Comment by mnaseri2770** [comment]
Thanks for your explanation, just one more thing, is it also possible to apply the noise directly to a model? I mean suppose I have trained my model on the training data and then to use it for test, I want to add some noise to it. Is it possible?

**Comment by ChrisWaites** [comment]
I'm not aware of such, as I believe a number of issues would arise if you were to attempt to do this in the context of differential privacy (although, I suppose you could technically view what you've described as equivalent to the case where you do exactly one step of optimization, which isn't very practical but on the table).
